spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java

```diff
@@ -192,6 +196,15 @@ TraceKeys traceKeys() {
 						}));
 	}
 
+	private static String getHttpRoute(ServerWebExchange exchange) {
```

Comment:
maybe make the notes about the corner cases as comments here (better as unit tests, but if no tests, then at least comments keeps the knowledge around)

I suspect that if you are in the `attribute instanceof HandlerMethod` block, then you'd know you are in a controller, so if the pattern string is null in this case, you'd coerse to empty to fit the contract.

Reply:
I am still convinced that coercion is not needed because pattern attribute is never null for MVC calls. For other calls returning empty string will overwrite span name even if user have already manually set custom one - can be the case with router functions.

I'll add some comments, extend existing test to check generated span name and let's call it a day, OK? :)

Reply:
to see what I mean, you can clone brave locally. add a test that makes the thing you are concerned about, for example, in webmvc. the current approach to span customization (for example if you have a better name) is to have your own parser.

https://github.com/openzipkin/brave/blob/master/instrumentation/http-tests/src/main/java/brave/test/http/ITHttpServer.java

Even if we dodge it here, if there is a problem, the trap is set elsewhere. Not trying to be arbitrary, just this sort of inconsistency limits the ability of a couple people to maintain projects that a lot of people use.

Reply:
About `HttpAdapter.route` result - I've read the docs again and it says "Null indicates this instrumentation doesn't understand http routes". `TraceWebFilter` does support routes so its implementation of HttpAdapter should never return null. Is that correct?

I feel like I've lost the direction of your though. What do you suggest?

Reply:
At least when "attribute instanceof HandlerMethod" I suspect we can coerce null to empty. I think making sure that this happens (if it is correct that attribute instanceof HandlerMethod when defined on an annotated class), then we make progress and keep things consistent. So, when you are inside the  "attribute instanceof HandlerMethod" if statement make sure getHttpRoute null coerces to empty somehow

There may be other cases, too, but we can defer that to later tests.

Make sense?

Independently, we need to put a subclass of ITHttpServer here, because it is better to show not tell. This is something I will try to do offline.

Reply:
OK, I redid the whole patch according to this. Also added a test for router function to demonstrate span naming behaviour.